File: server/apps/lessons/application/dto/create_subject_lesson_plan_dto.py
from dataclasses import dataclass
from typing import Optional


@dataclass
class CreateSubjectLessonPlanRuleDTO:

    lesson_type: str

    min_credits: Optional[int]

    max_credits: Optional[int]

    total_lessons: int

    @classmethod
    def from_dict(
        cls,
        data: dict,
    ) -> "CreateSubjectLessonPlanRuleDTO":

        return cls(
            lesson_type=data["lesson_type"],
            min_credits=data.get("min_credits"),
            max_credits=data.get("max_credits"),
            total_lessons=data["total_lessons"],
        )


@dataclass
class CreateSubjectLessonPlanDTO:

    semester_number: str

    rules: list[CreateSubjectLessonPlanRuleDTO]

    @classmethod
    def from_dict(
        cls,
        data: dict,
    ) -> "CreateSubjectLessonPlanDTO":

        rules = [
            CreateSubjectLessonPlanRuleDTO.from_dict(
                rule
            )
            for rule in data["rules"]
        ]

        return cls(
            semester_number=data["semester_number"],
            rules=rules,
        )

File: client/src/application/lessons/dto/create_subject_lesson_plan.dto.js
class CreateSubjectLessonPlanRuleDTO {
  constructor(data = {}) {
    this.lessonType = data.lessonType;

    this.minCredits = data.minCredits ?? null;

    this.maxCredits = data.maxCredits ?? null;

    this.totalLessons = data.totalLessons;
  }

  toJSON() {
    return {
      lesson_type: this.lessonType,

      min_credits: this.minCredits,

      max_credits: this.maxCredits,

      total_lessons: this.totalLessons,
    };
  }
}

class CreateSubjectLessonPlanDTO {
  constructor(data = {}) {
    this.semesterNumber = data.semesterNumber;

    this.rules = (data.rules || []).map((rule) =>
      rule instanceof CreateSubjectLessonPlanRuleDTO
        ? rule
        : new CreateSubjectLessonPlanRuleDTO(rule),
    );
  }

  static fromRequestBody(body = {}) {
    const rules = [];

    const source = body.rules || {};

    for (const lessonType of ["theory", "practice"]) {
      const typeRules = source[lessonType] || {};

      const items = Array.isArray(typeRules)
        ? typeRules
        : Object.values(typeRules);

      for (const rule of items) {
        rules.push(
          new CreateSubjectLessonPlanRuleDTO({
            lessonType,

            minCredits:
              rule.min_credits !== undefined && rule.min_credits !== ""
                ? Number(rule.min_credits)
                : null,

            maxCredits:
              rule.max_credits !== undefined && rule.max_credits !== ""
                ? Number(rule.max_credits)
                : null,

            totalLessons:
              rule.total_lessons !== undefined && rule.total_lessons !== ""
                ? Number(rule.total_lessons)
                : null,
          }),
        );
      }
    }

    return new CreateSubjectLessonPlanDTO({
      semesterNumber: body.semester_number,

      rules,
    });
  }

  toJSON() {
    return {
      semester_number: this.semesterNumber,

      rules: this.rules.map((rule) => rule.toJSON()),
    };
  }
}

module.exports = CreateSubjectLessonPlanDTO;
